File: programs/sdk/assemblyscript/assembly/ids.ts
/**
 * Exact, nonzero protocol identifiers.
 *
 * Storage is private and every public byte view is a fresh exact-width copy,
 * so callers cannot mutate an identifier after validation or hand a host call
 * a short array whose pointer is advertised as thirty-two bytes.
 */

import { IDENTIFIER_BYTES, identifierFromWords, isZero, slice } from "./bytes";

function canonical(source: StaticArray<u8>, offset: i32): StaticArray<u8> | null {
  if (offset < 0 || offset > source.length - IDENTIFIER_BYTES) return null;
  const value = slice(source, offset, IDENTIFIER_BYTES);
  return isZero(value) ? null : value;
}

function clone(value: StaticArray<u8>): StaticArray<u8> {
  return slice(value, 0, IDENTIFIER_BYTES);
}

/** Stable identifier of a deployed program. */
export class ProgramId {
  private readonly value: StaticArray<u8>;

  private constructor(value: StaticArray<u8>) { this.value = value; }

  static fromWords(word0: u64, word1: u64, word2: u64, word3: u64): ProgramId | null {
    return ProgramId.fromBytes(identifierFromWords(word0, word1, word2, word3), 0);
  }

  static fromBytes(source: StaticArray<u8>, offset: i32): ProgramId | null {
    const value = canonical(source, offset);
    return value === null ? null : new ProgramId(value);
  }

  get bytes(): StaticArray<u8> { return clone(this.value); }
  isReserved(): bool { return false; }
}

/** Stable identifier of a protocol asset. */
export class AssetId {
  private readonly value: StaticArray<u8>;

  private constructor(value: StaticArray<u8>) { this.value = value; }

  static fromWords(word0: u64, word1: u64, word2: u64, word3: u64): AssetId | null {
    return AssetId.fromBytes(identifierFromWords(word0, word1, word2, word3), 0);
  }

  static fromBytes(source: StaticArray<u8>, offset: i32): AssetId | null {
    const value = canonical(source, offset);
    return value === null ? null : new AssetId(value);
  }

  get bytes(): StaticArray<u8> { return clone(this.value); }
  isReserved(): bool { return false; }
}

/** Stable identifier of an account a 402LXP transfer may credit. */
export class AccountId {
  private readonly value: StaticArray<u8>;

  private constructor(value: StaticArray<u8>) { this.value = value; }

  static fromWords(word0: u64, word1: u64, word2: u64, word3: u64): AccountId | null {
    return AccountId.fromBytes(identifierFromWords(word0, word1, word2, word3), 0);
  }

  static fromBytes(source: StaticArray<u8>, offset: i32): AccountId | null {
    const value = canonical(source, offset);
    return value === null ? null : new AccountId(value);
  }

  get bytes(): StaticArray<u8> { return clone(this.value); }
  isReserved(): bool { return false; }
}

/** Digest naming one verified protocol receipt. */
export class ReceiptDigest {
  private readonly value: StaticArray<u8>;

  private constructor(value: StaticArray<u8>) { this.value = value; }

  static fromWords(word0: u64, word1: u64, word2: u64, word3: u64): ReceiptDigest | null {
    return ReceiptDigest.fromBytes(identifierFromWords(word0, word1, word2, word3), 0);
  }

  static fromBytes(source: StaticArray<u8>, offset: i32): ReceiptDigest | null {
    const value = canonical(source, offset);
    return value === null ? null : new ReceiptDigest(value);
  }

  get bytes(): StaticArray<u8> { return clone(this.value); }
  isReserved(): bool { return false; }
}

File: programs/sdk/rust/src/ids.rs
//! Nonzero protocol identifiers.
//!
//! Every identifier the ABI carries is 32 bytes wide and reserves the all-zero
//! value for absence. Constructing one checks that bound, so a program cannot
//! hand the host an identifier the runtime would have to refuse.

use crate::error::{Field, ProgramError, Reason};

const IDENTIFIER_BYTES: usize = 32;

const fn is_reserved(bytes: &[u8; IDENTIFIER_BYTES]) -> bool {
    let mut index = 0;
    while index < IDENTIFIER_BYTES {
        if bytes[index] != 0 {
            return false;
        }
        index += 1;
    }
    true
}

/// Stable identifier of a deployed program.
#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct ProgramId([u8; IDENTIFIER_BYTES]);

impl ProgramId {
    /// Constructs a nonzero program identifier.
    ///
    /// # Errors
    ///
    /// Refuses the all-zero identifier reserved for absence.
    pub const fn new(bytes: [u8; IDENTIFIER_BYTES]) -> Result<Self, ProgramError> {
        if is_reserved(&bytes) {
            return Err(ProgramError::value(Field::Program, Reason::Zero));
        }
        Ok(Self(bytes))
    }

    /// Returns the canonical identifier bytes.
    #[must_use]
    pub const fn bytes(self) -> [u8; IDENTIFIER_BYTES] {
        self.0
    }
}

/// Stable identifier of a protocol asset.
#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct AssetId([u8; IDENTIFIER_BYTES]);

impl AssetId {
    /// Constructs a nonzero asset identifier.
    ///
    /// # Errors
    ///
    /// Refuses the all-zero identifier reserved for absence.
    pub const fn new(bytes: [u8; IDENTIFIER_BYTES]) -> Result<Self, ProgramError> {
        if is_reserved(&bytes) {
            return Err(ProgramError::value(Field::Asset, Reason::Zero));
        }
        Ok(Self(bytes))
    }

    /// Returns the canonical identifier bytes.
    #[must_use]
    pub const fn bytes(self) -> [u8; IDENTIFIER_BYTES] {
        self.0
    }
}

/// Stable identifier of an account a 402LXP transfer may credit.
#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct AccountId([u8; IDENTIFIER_BYTES]);

impl AccountId {
    /// Constructs a nonzero account identifier.
    ///
    /// # Errors
    ///
    /// Refuses the all-zero identifier reserved for absence.
    pub const fn new(bytes: [u8; IDENTIFIER_BYTES]) -> Result<Self, ProgramError> {
        if is_reserved(&bytes) {
            return Err(ProgramError::value(Field::Account, Reason::Zero));
        }
        Ok(Self(bytes))
    }

    /// Returns the canonical identifier bytes.
    #[must_use]
    pub const fn bytes(self) -> [u8; IDENTIFIER_BYTES] {
        self.0
    }
}

/// Digest naming one verified protocol receipt.
#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct ReceiptDigest([u8; IDENTIFIER_BYTES]);

impl ReceiptDigest {
    /// Constructs a nonzero receipt digest.
    ///
    /// # Errors
    ///
    /// Refuses the all-zero digest reserved for absence.
    pub const fn new(bytes: [u8; IDENTIFIER_BYTES]) -> Result<Self, ProgramError> {
        if is_reserved(&bytes) {
            return Err(ProgramError::value(Field::ReceiptDigest, Reason::Zero));
        }
        Ok(Self(bytes))
    }

    /// Returns the canonical digest bytes.
    #[must_use]
    pub const fn bytes(self) -> [u8; IDENTIFIER_BYTES] {
        self.0
    }
}
